### ISSUE TITLE
Consuming Swift packages in projects directly or through Dependencies.swift results in different behaviour

### ISSUE THREAD

**Issue by danibachar** [tuist/tuist#5784]
### What problem or need do you have?

A Swift Package code working fine when working on on a project that consumes the packages directly. 
Fails when consuming the packages as `external` package using Dependencies.swift

A full description of the issue and an example project with reproducible steps can be found in [this repo](https://github.com/danibachar/TuistTypeCastInModules)

### Potential solution

The workaround I found is also described in the repo readme file.
Which was to extract the code from one of the Swift Packages and have it resides at the app level

### macOS version

Ventura 13.4 (22F66)

### Tuist version

3.37.0

### Xcode version

15.1

**Comment by pepicrft** [comment]
Thanks for reporting this @danibachar and providing a reproducible example. We'll try to fix the issue.

**Comment by danibachar** [comment]
I wanted to update, on a temporary fix I found.
Added this [PR](https://github.com/danibachar/TuistTypeCastInModules/pull/1) to my main branch
TL;DR - settings the problematic Package to a `framework` or `.dynamicLibrary` solved the issue.
The problem that I left with is that I need all the related tree to be dynamically linked as static+dynamic combination cause Tuist warning and can basically cause duplicated symbols

I'd love to learn about this issue and understand how/if Tuist plans to adapt to it?

**Comment by pepicrft** [comment]
@danibachar Tuist defaults to static libraries when the package specifies that the linking is "automatic". Xcode build system resolves that linking automatically when using the standard SPM integration. However,  Tuist relies on you deciding whether you'd like your packages to be frameworks or libraries, and static or dynamic. Which one is the most suitable depends on the graph. We could provide guidelines, but I don't think Tuist should magically make that decision on your behalf because it has implications that we believe developers should be aware of and be conscious about.

**Comment by danibachar** [comment]
Thanks for the feedback @pepicrft - I understand that its hard (or maybe unwanted in Tuist?) to mimic Xcode heuristics around that - it was a bit unexpected to see different behaviour around that - so I wanted to point that out. Now I understand the reason (default to static linking).

The configuration in Dependencies.swift - which I hope will be migrated and supported (i think you mentioned in the past it will through the Config/Worksapce files?) in Tuist 4 gives us a good enough workaround that issue.

Thanks again for your comment.

**Comment by pepicrft** [comment]
> Thanks for the feedback @pepicrft - I understand that its hard (or maybe unwanted in Tuist?) to mimic Xcode heuristics around that - it was a bit unexpected to see different behaviour around that - so I wanted to point that out. Now I understand the reason (default to static linking).

My pleasure! Note that it's not about difficulty but about optimizability (if this is even a word). Your project goes through this pipeline:
```
Tuist project - [generation] ->  Xcode project -> [build] -> Binaries
```
Xcode projects support implicit and ambiguous configuration that is resolved at build time. What is automatic? What depends on what? It's fine! Let us figure that out, Xcode says. While that is good convenience for small-scale projects, is a terrible idea for large-scale ones because it might introduce non-determinism and brittleness.

Tuist leverages the pre-build generation phase to disallow that pattern. Instead, we require things to be explicit with some defaults that we believe are sensible. In this case the default, which is static across all Tuist projects, deviates from the value that you get in Xcode, which is dynamically resolved at built time and might vary without you controlling it.

> Thanks again for your comment. I think this issue can be closed (might be worth documenting - maybe in a discussion or the new website? If you think so - please direct me where I can share this knowledge)

We have a [dependencies](https://docs.tuist.io/documentation/tuist/dependencies) page in the documentation, so I think that's a good place to document what's Tuist's stance when it comes to this.

**Comment by dainelbac** [comment]
I wanted to update that I tried adding the `GENERATE_MASTER_OBJECT_FILE=true` flag and it didn't solve the issue.
I run into [this SO](https://stackoverflow.com/questions/3276474/symbol-hiding-in-static-libraries-built-with-xcode) thread about static vs dynamic linkage difference which I find interesting

**Comment by danibachar** [comment]
I wanted to update that I tried adding the `GENERATE_MASTER_OBJECT_FILE=true` flag and it didn't solve the issue.
I run into [this SO](https://stackoverflow.com/questions/3276474/symbol-hiding-in-static-libraries-built-with-xcode) thread about static vs dynamic linkage difference which I find interesting
